x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityPainlessExecuteIT.java

```diff
@@ -0,0 +1,322 @@
+    public void testPainlessExecute() throws Exception {
+        configureRemoteCluster();
+
+        {
+            // Query cluster -> add role for test user - do not give any privileges for remote_indices
+            final var putRoleRequest = new Request("PUT", "/_security/role/" + REMOTE_SEARCH_ROLE);
+            putRoleRequest.setJsonEntity("""
+                {
+                  "indices": [
+                    {
+                      "names": ["local_index", "my_local*"],
+                      "privileges": ["read"]
+                    }
+                  ]
+                }""");
+            assertOK(adminClient().performRequest(putRoleRequest));
+
+            // Query cluster -> create user and assign role
+            final var putUserRequest = new Request("PUT", "/_security/user/" + REMOTE_SEARCH_USER);
+            putUserRequest.setJsonEntity("""
+                {
+                  "password": "x-pack-test-password",
+                  "roles" : ["remote_search"]
+                }""");
+            assertOK(adminClient().performRequest(putUserRequest));
+
+            // Query cluster -> create test index
+            final var indexDocRequest = new Request("POST", "/local_index/_doc?refresh=true");
+            indexDocRequest.setJsonEntity("{\"local_foo\": \"local_bar\"}");
+            assertOK(client().performRequest(indexDocRequest));
+
+            // Fulfilling cluster -> create test indices
+            final Request bulkRequest = new Request("POST", "/_bulk?refresh=true");
+            bulkRequest.setJsonEntity(Strings.format("""
+                { "index": { "_index": "index1" } }
+                { "foo": "bar" }
+                { "index": { "_index": "secretindex" } }
+                { "bar": "foo" }
+                """));
+            assertOK(performRequestAgainstFulfillingCluster(bulkRequest));
+        }
+
+        {
+            // TEST CASE 1: Query local cluster for local_index - should work since role has read perms for it
+            Request painlessExecuteLocal = createPainlessExecuteRequest("local_index");
+
+            Response response = performRequestWithRemoteSearchUser(painlessExecuteLocal);
+            assertOK(response);
+            String responseBody = EntityUtils.toString(response.getEntity());
+            assertThat(responseBody, equalTo("{\"result\":[\"test\"]}"));
+        }
+        {
+//            // TEST CASE 2: Query remote cluster for index1 - should fail since no permissions granted for remote clusters yet
+//            Request painlessExecuteRemote = createPainlessExecuteRequest("my_remote_cluster:index1");
+//            Exception exc = expectThrows(Exception.class, () -> performRequestWithRemoteSearchUser(painlessExecuteRemote));
+//            assertThat(exc, notNullValue());
+//            assertThat(exc.getMessage(), containsString("Connection is closed"));  // TODO: is this the expected exception?
```

Comment:
@jakelandis  - if I uncomment this test case, it receives a "ConnectionClosed" exception rather than a 403 security exception. And worse, the next usage of the client to call to the server (line 219) below, also fails with a ConnectionClosed exception. Any idea why this occurs? 

[Here](https://gist.github.com/quux00/8e5a6f8847f049b23880f44a32643511) is the detailed logs output from the IT test.

You can see there that buried in there is an auth originated error:

```
Caused by: java.lang.AssertionError
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationDenialMessages.remoteActionDenied(AuthorizationDenialMessages.java:95)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$remoteActionDenied$31(AuthorizationService.java:935)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.denialException(AuthorizationService.java:988)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.remoteActionDenied(AuthorizationService.java:932)
```

but it seems to cause a fatal error at the Transport layer that seems problematic. I wasn't able to reproduce this problem on my manual testing with a 1.x security set up.

<br/>

And **STDERR output is**:

```
XXX Querying for [local_index]
XXX Querying for [my_remote_cluster:index1]
XXX >>> ABOUT TO UPDATE ROLES

ERROR: Elasticsearch exited unexpectedly, with exit code 1
```

So you can see it fails upon trying to update the roles.

Reply:
The `ConnectionClosed` exception isn't the problem, that's what we'd expect when a node trips an assertion. But tripping an assertion is most definitely a problem. Without having investigated at all, I would guess it's because `cluster:admin/scripts/painless/execute` does not start with `indices:`, but that prefix is required for actions that involve indices and authorization.

Reply:
This same scenario works (returns the expected security_exception) in the RCS1.0 based IT test, but still fails (with the Assertion/ConnectionClosed issue) in the RCS2.0 based IT test, so I plan to follow up with Jake (or someone on the es-security side) to try to get to root cause about what is happening with RCS2.0. It could be test set-up or a bug somewhere in the system.

Reply:
FTR I'm _very_ concerned about this PR having been merged without finding out more about this tripping assertion. From the information I've seen above, I think the most likely explanation is that that this PR itself introduces a bug (or at least violates some assumptions that the security module makes, with currently-unknown consequences).

Reply:
This indeed is an issue with `cluster:admin/scripts/painless/execute` action being the cluster action (and not indices). It breaks the assumption for RCS 2.0 security, which allows only indices actions to be granted to remote clusters. 

Looking at the code, the painless action seems to be "special" in a sense that it gets [authorized as the indices action](https://github.com/elastic/elasticsearch/pull/85512) when index name is specified.

https://github.com/elastic/elasticsearch/blob/81332093550e6f11b1ecca1ddae345cce06d9c86/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/SecurityActionMapper.java#L46-L51

---

This test is hitting an edge case where request gets [rejected on a local cluster](https://github.com/elastic/elasticsearch/pull/93807) when "user" has no remote index privileges. This is an optimization which avoids sending a request to remote cluster that will anyway be rejected. The assertion is being triggered because this "fail early" approach is executed before `SecurityActionMapper` does the replacement.

I don't see this as a security issue since the request will be rejected with a proper access denied message. But nevertheless, we should handle this case as well.

https://github.com/elastic/elasticsearch/blob/44b0047db5ed516c92a9f5d952319b17f29047c8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java#L95-L99

Reply:
++ thanks for investigating further Slobodan. I'm now less concerned about this specific case, but remain very concerned about the sequence of events that led to this situation. We might not be so lucky next time.